File: Autonomous_Mobile_Robots_Course/Chapter3/Lesson2/Chapter3_Lesson2.m
% Chapter 3 - Lesson 2: Curvature and Turning Radius Limits (MATLAB/Simulink)
%
% This script:
%   1) Computes discrete curvature along a sampled 2D path.
%   2) Converts curvature to bicycle steering angle delta = atan(L*kappa).
%   3) Checks steering limits and lateral-acceleration speed limits v_max(s).
%   4) (Optional) Builds a minimal Simulink bicycle-kinematics model programmatically.
%
% Recommended toolboxes (optional):
%   - Robotics System Toolbox (for integration with mobile robot workflows)
%   - Simulink (for model generation)
%
% Author: Course generator

clear; clc;

%% Parameters
L = 0.35;                      % wheelbase [m]
delta_max = deg2rad(28);        % steering angle limit [rad]
a_lat_max = 2.0;                % lateral acceleration limit [m/s^2]

%% Example path (students may replace)
N = 401;
s = linspace(0, 12, N);
x = s;
y = 1.2*sin(0.7*s) + 0.2*sin(2.2*s);
P = [x(:), y(:)];

%% Discrete curvature using 3-point circumcircle formula
kappa = zeros(N,1);
for i = 2:N-1
    p0 = P(i-1,:);
    p1 = P(i,:);
    p2 = P(i+1,:);

    v01 = p1 - p0;
    v02 = p2 - p0;
    v12 = p2 - p1;

    a = norm(v12);
    b = norm(v02);
    c = norm(v01);

    area2 = v01(1)*v02(2) - v01(2)*v02(1); % signed double area

    denom = a*b*c;
    if denom < 1e-12
        kappa(i) = 0;
    else
        kappa(i) = 2*area2/denom;
    end
end

%% Bicycle steering mapping and limits
delta = atan(L*kappa);
kappa_max_steer = tan(delta_max)/L;

steer_ok = abs(delta) <= delta_max + 1e-12;
fprintf('kappa_max_steer = %.4f 1/m\n', kappa_max_steer);
fprintf('Peak |kappa|    = %.4f 1/m\n', max(abs(kappa)));
fprintf('Steering-feasible? %s\n', string(all(abs(kappa) <= kappa_max_steer + 1e-12)));

%% Lateral acceleration speed limits: v <= sqrt(a_lat_max/|kappa|)
vmax = inf(N,1);
mask = abs(kappa) > 1e-12;
vmax(mask) = sqrt(a_lat_max ./ abs(kappa(mask)));

%% Visualizations
figure; plot(x,y,'LineWidth',1.5); axis equal; grid on;
title('Path (x,y)'); xlabel('x [m]'); ylabel('y [m]');

figure; plot(s,kappa,'LineWidth',1.5); grid on;
title('Curvature kappa(s)'); xlabel('s (index)'); ylabel('kappa [1/m]');

figure; plot(s,vmax,'LineWidth',1.5); grid on;
title('Speed upper bound from lateral acceleration'); xlabel('s (index)'); ylabel('v_{max} [m/s]');

%% Optional: Build a simple Simulink bicycle kinematics model
% This creates a model with states (x,y,theta) and inputs (v, delta).
% To run it:
%   - Ensure Simulink is installed.
%   - Uncomment the lines below.
%
% modelName = 'Chapter3_Lesson2_BicycleModel';
% buildSimulinkBicycleModel(modelName, L);
% open_system(modelName);

function buildSimulinkBicycleModel(modelName, L)
    if bdIsLoaded(modelName)
        close_system(modelName, 0);
    end
    new_system(modelName);
    open_system(modelName);

    % Blocks
    add_block('simulink/Sources/In1', [modelName '/v']);
    add_block('simulink/Sources/In1', [modelName '/delta']);
    add_block('simulink/Math Operations/Trigonometric Function', [modelName '/cos'], 'Operator', 'cos');
    add_block('simulink/Math Operations/Trigonometric Function', [modelName '/sin'], 'Operator', 'sin');
    add_block('simulink/Math Operations/Trigonometric Function', [modelName '/tan'], 'Operator', 'tan');

    add_block('simulink/Math Operations/Product', [modelName '/vcos']);
    add_block('simulink/Math Operations/Product', [modelName '/vsin']);
    add_block('simulink/Math Operations/Product', [modelName '/v_tan']);
    add_block('simulink/Math Operations/Gain', [modelName '/1_over_L'], 'Gain', num2str(1/L));

    add_block('simulink/Continuous/Integrator', [modelName '/Int_x']);
    add_block('simulink/Continuous/Integrator', [modelName '/Int_y']);
    add_block('simulink/Continuous/Integrator', [modelName '/Int_theta']);

    add_block('simulink/Sinks/Out1', [modelName '/x']);
    add_block('simulink/Sinks/Out1', [modelName '/y']);
    add_block('simulink/Sinks/Out1', [modelName '/theta']);

    % Connections
    add_line(modelName, 'delta/1', 'tan/1');
    add_line(modelName, 'tan/1', 'v_tan/2');
    add_line(modelName, 'v/1', 'v_tan/1');
    add_line(modelName, 'v_tan/1', '1_over_L/1');
    add_line(modelName, '1_over_L/1', 'Int_theta/1');
    add_line(modelName, 'Int_theta/1', 'cos/1');
    add_line(modelName, 'Int_theta/1', 'sin/1');
    add_line(modelName, 'v/1', 'vcos/1');
    add_line(modelName, 'cos/1', 'vcos/2');
    add_line(modelName, 'v/1', 'vsin/1');
    add_line(modelName, 'sin/1', 'vsin/2');
    add_line(modelName, 'vcos/1', 'Int_x/1');
    add_line(modelName, 'vsin/1', 'Int_y/1');

    add_line(modelName, 'Int_x/1', 'x/1');
    add_line(modelName, 'Int_y/1', 'y/1');
    add_line(modelName, 'Int_theta/1', 'theta/1');

    set_param(modelName, 'StopTime', '10');
    save_system(modelName);
end

File: Autonomous_Mobile_Robots_Course/Chapter3/Lesson2/Chapter3_Lesson2.cpp
/*
Chapter 3 - Lesson 2: Curvature and Turning Radius Limits (C++)

Build note:
  This example uses Eigen (header-only). If you do not have Eigen,
  you can replace Eigen::Vector2d with a small custom struct.

Example compilation (Linux/macOS):
  g++ -O2 -std=c++17 Chapter3_Lesson2.cpp -I /usr/include/eigen3 -o Chapter3_Lesson2

This program:
  1) Computes discrete signed curvature along a polyline using a 3-point circumcircle formula.
  2) Checks steering curvature limit kappa_max = tan(delta_max)/L.
*/

#include <iostream>
#include <vector>
#include <cmath>
#include <Eigen/Dense>

using Vec2 = Eigen::Vector2d;

static double signed_curvature_three_points(const Vec2& p0, const Vec2& p1, const Vec2& p2) {
    Vec2 v01 = p1 - p0;
    Vec2 v02 = p2 - p0;
    Vec2 v12 = p2 - p1;

    double a = v12.norm();
    double b = v02.norm();
    double c = v01.norm();

    // Signed double area (2*A) in 2D: cross(v01, v02)
    double area2 = v01.x() * v02.y() - v01.y() * v02.x();

    double denom = a * b * c;
    if (denom < 1e-12) return 0.0;
    return 2.0 * area2 / denom; // since kappa = 4*A/(abc) and area2 = 2*A
}

static std::vector<double> curvature_polyline(const std::vector<Vec2>& P) {
    const int N = static_cast<int>(P.size());
    std::vector<double> kappa(N, 0.0);
    for (int i = 1; i < N - 1; ++i) {
        kappa[i] = signed_curvature_three_points(P[i-1], P[i], P[i+1]);
    }
    return kappa;
}

static double kappa_max_steer(double L, double delta_max_rad) {
    return std::tan(delta_max_rad) / L;
}

int main() {
    // Create a sample path
    const int N = 401;
    std::vector<Vec2> P;
    P.reserve(N);
    for (int i = 0; i < N; ++i) {
        double s = 12.0 * static_cast<double>(i) / static_cast<double>(N - 1);
        double x = s;
        double y = 1.2 * std::sin(0.7 * s) + 0.2 * std::sin(2.2 * s);
        P.emplace_back(x, y);
    }

    auto kappa = curvature_polyline(P);

    // Bicycle steering limit
    double L = 0.35;
    double delta_max = 28.0 * M_PI / 180.0;
    double kappa_max = kappa_max_steer(L, delta_max);

    bool ok = true;
    double kappa_peak = 0.0;
    for (double ki : kappa) {
        kappa_peak = std::max(kappa_peak, std::abs(ki));
        if (std::abs(ki) > kappa_max + 1e-12) ok = false;
    }

    std::cout << "kappa_max_steer = " << kappa_max << " 1/m\n";
    std::cout << "peak |kappa|    = " << kappa_peak << " 1/m\n";
    std::cout << "steering-feasible? " << (ok ? "YES" : "NO") << "\n";
    return 0;
}
